File: go/internal/ui/toggle.go
package ui

import (
	"fmt"
	"strings"
)

// LabelPosition represents the available label position options for the toggle
type LabelPosition string

const (
	LabelPositionTop    LabelPosition = "top"
	LabelPositionLeft   LabelPosition = "left"
	LabelPositionHidden LabelPosition = "hidden"
)

// Toggle represents a toggle component
type Toggle struct {
	Classes       string
	Label         string
	LabelPosition LabelPosition
	ID            string
	Name          string
	Help          string
	Data          map[string]interface{}
	IsDisabled    bool
	IsChecked     bool
	IsLoading     bool
	Content       string
}

// NewToggle creates a new toggle with the given parameters
func NewToggle(
	classes string,
	label string,
	labelPosition LabelPosition,
	id string,
	name string,
	help string,
	data map[string]interface{},
	isDisabled bool,
	isChecked bool,
	isLoading bool,
	content string,
) *Toggle {
	if data == nil {
		data = make(map[string]interface{})
	}

	return &Toggle{
		Classes:       classes,
		Label:         label,
		LabelPosition: labelPosition,
		ID:            id,
		Name:          name,
		Help:          help,
		Data:          data,
		IsDisabled:    isDisabled,
		IsChecked:     isChecked,
		IsLoading:     isLoading,
		Content:       content,
	}
}

// GetDataAttributes returns the data attributes for the toggle
func (t *Toggle) GetDataAttributes() string {
	data := map[string]interface{}{
		"name":           t.Name,
		"id":             t.ID,
		"is_checked":     fmt.Sprintf("%v", t.IsChecked),
		"disabled":       fmt.Sprintf("%v", t.IsDisabled),
		"is_loading":     fmt.Sprintf("%v", t.IsLoading),
		"label":          t.Label,
		"help":           t.Help,
		"label_position": t.LabelPosition,
	}

	// Merge with additional data attributes
	for k, v := range t.Data {
		data[k] = v
	}

	return formatAttributes(data)
}

// Render generates the HTML for the toggle
func (t *Toggle) Render() string {
	var parts []string

	// Start toggle span
	parts = append(parts, fmt.Sprintf(`<span class="%s" %s>`, t.Classes, t.GetDataAttributes()))

	// Add help label if content is present
	if t.Content != "" {
		parts = append(parts, fmt.Sprintf(`<div class="gl-help-label">%s</div>`, t.Content))
	}

	// End toggle span
	parts = append(parts, "</span>")

	return strings.Join(parts, "\n")
}

File: rust_app/src/components/pajamas/toggle.rs
use serde::{Deserialize, Serialize};
use std::collections::HashMap;

/// Label position options for the toggle component
#[derive(Debug, Clone, Copy, PartialEq, Serialize, Deserialize)]
pub enum LabelPosition {
    Top,
    Left,
    Hidden,
}

impl LabelPosition {
    fn as_str(&self) -> &'static str {
        match self {
            LabelPosition::Top => "top",
            LabelPosition::Left => "left",
            LabelPosition::Hidden => "hidden",
        }
    }
}

/// Renders a GlToggle root element
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ToggleComponent {
    /// CSS classes
    classes: String,
    /// Label text
    label: Option<String>,
    /// Label position
    label_position: Option<LabelPosition>,
    /// HTML ID
    id: Option<String>,
    /// Input name
    name: Option<String>,
    /// Help text
    help: Option<String>,
    /// Data attributes
    data: HashMap<String, String>,
    /// Whether the toggle is disabled
    is_disabled: bool,
    /// Whether the toggle is checked
    is_checked: bool,
    /// Whether the toggle is loading
    is_loading: bool,
    /// Rich help text content
    content: Option<String>,
}

impl ToggleComponent {
    /// Create a new toggle component
    pub fn new(
        classes: String,
        label: Option<String>,
        label_position: Option<LabelPosition>,
        id: Option<String>,
        name: Option<String>,
        help: Option<String>,
        data: Option<HashMap<String, String>>,
        is_disabled: bool,
        is_checked: bool,
        is_loading: bool,
        content: Option<String>,
    ) -> Self {
        Self {
            classes,
            label,
            label_position,
            id,
            name,
            help,
            data: data.unwrap_or_default(),
            is_disabled,
            is_checked,
            is_loading,
            content,
        }
    }

    /// Render the toggle component
    pub fn render(&self) -> String {
        let mut html = format!(
            r#"<span class="{}"{}>"#,
            self.classes,
            self.render_data_attributes()
        );

        // Render rich help text content if present
        if let Some(content) = &self.content {
            html.push_str(&format!(r#"<div class="gl-help-label">{}</div>"#, content));
        }

        html.push_str("</span>");
        html
    }

    /// Render data attributes
    fn render_data_attributes(&self) -> String {
        let mut data_attrs = self.data.clone();

        // Add standard data attributes
        if let Some(name) = &self.name {
            data_attrs.insert("name".to_string(), name.clone());
        }
        if let Some(id) = &self.id {
            data_attrs.insert("id".to_string(), id.clone());
        }
        if let Some(label) = &self.label {
            data_attrs.insert("label".to_string(), label.clone());
        }
        if let Some(help) = &self.help {
            data_attrs.insert("help".to_string(), help.clone());
        }
        if let Some(label_position) = &self.label_position {
            data_attrs.insert(
                "label-position".to_string(),
                label_position.as_str().to_string(),
            );
        }

        // Add boolean attributes
        data_attrs.insert("is-checked".to_string(), self.is_checked.to_string());
        data_attrs.insert("disabled".to_string(), self.is_disabled.to_string());
        data_attrs.insert("is-loading".to_string(), self.is_loading.to_string());

        // Format data attributes
        data_attrs
            .iter()
            .map(|(key, value)| format!(" data-{}={}", key, value))
            .collect()
    }
}
